test/Conversion/SCFToCalyx/convert_func.mlir

```diff
@@ -0,0 +1,304 @@
+// CHECK-NEXT:   calyx.wires {
+// CHECK-NEXT:     calyx.assign %out0 = %ret_arg0_reg.out : i32
+// CHECK-NEXT:     calyx.group @init_func_instance {
+// CHECK-NEXT:       calyx.assign %func_instance.reset = %true : i1
+// CHECK-NEXT:       calyx.assign %func_instance.go = %true : i1
+// CHECK-NEXT:       calyx.group_done %func_instance.done : i1
+// CHECK-NEXT:     }
+// CHECK-NEXT:     calyx.comb_group @bb0_2 {
+// CHECK-NEXT:       calyx.assign %std_ge_0.left = %std_add_1.out : i32
+// CHECK-NEXT:       calyx.assign %std_ge_0.right = %in2 : i32
+// CHECK-NEXT:       calyx.assign %std_add_1.left = %std_add_0.out : i32
+// CHECK-NEXT:       calyx.assign %std_add_0.left = %in0 : i32
+// CHECK-NEXT:       calyx.assign %std_add_0.right = %in1 : i32
+// CHECK-NEXT:       calyx.assign %std_add_1.right = %in1 : i32
+// CHECK-NEXT:     }
+// CHECK-NEXT:     calyx.group @ret_assign_0 {
+// CHECK-NEXT:       calyx.assign %ret_arg0_reg.in = %func_instance.out0 : i32
+// CHECK-NEXT:       calyx.assign %ret_arg0_reg.write_en = %true : i1
+// CHECK-NEXT:       calyx.group_done %ret_arg0_reg.done : i1
+// CHECK-NEXT:     }
+// CHECK-NEXT:     calyx.group @ret_assign_1 {
+// CHECK-NEXT:       calyx.assign %ret_arg0_reg.in = %func_instance.out0 : i32
+// CHECK-NEXT:       calyx.assign %ret_arg0_reg.write_en = %true : i1
+// CHECK-NEXT:       calyx.group_done %ret_arg0_reg.done : i1
+// CHECK-NEXT:     }
+// CHECK-NEXT:   }
+// CHECK-NEXT:   calyx.control {
+// CHECK-NEXT:     calyx.seq {
+// CHECK-NEXT:       calyx.if %std_ge_0.out with @bb0_2 {
+// CHECK-NEXT:         calyx.seq {
+// CHECK-NEXT:           calyx.enable @init_func_instance
+// CHECK-NEXT:           calyx.invoke @func_instance(%func_instance.in0 = %std_add_0.out, %func_instance.in1 = %std_add_1.out) -> (i32, i32)
+// CHECK-NEXT:           calyx.enable @ret_assign_0
+// CHECK-NEXT:         }
+// CHECK-NEXT:       } else {
+// CHECK-NEXT:         calyx.seq {
+// CHECK-NEXT:           calyx.enable @init_func_instance
+// CHECK-NEXT:           calyx.invoke @func_instance(%func_instance.in0 = %std_add_0.out, %func_instance.in1 = %std_add_1.out) -> (i32, i32)
+// CHECK-NEXT:           calyx.enable @ret_assign_1
+// CHECK-NEXT:         }
+// CHECK-NEXT:       }
+// CHECK-NEXT:     }
+// CHECK-NEXT:   }
+// CHECK-NEXT: } {toplevel}
+
+module {
+  func.func @func(%a0 : i32, %a1 : i32) -> i32 {
+    %0 = arith.addi %a0, %a1 : i32
+    %1 = arith.shli %0, %a0 : i32
+    %2 = arith.subi %1, %0 : i32
+    return %2 : i32
+  }
+
+  func.func @main(%a0 : i32, %a1 : i32, %a2 : i32) -> i32 {
+    %0 = arith.addi %a0, %a1 : i32
+    %1 = arith.addi %0, %a1 : i32
+    %b = arith.cmpi uge, %1, %a2 : i32
+    cf.cond_br %b, ^bb1, ^bb2
+  ^bb1:
+    %ret0 = func.call @func(%0, %1) : (i32, i32) -> i32
+    return %ret0 : i32
+  ^bb2:
+    %ret1 = func.call @func(%0, %1) : (i32, i32) -> i32
+    return %ret1 : i32
```

Comment:
Generally wrt. the tests in this file - i don't think that it's necessary to have a check on the entire output IR - probably just the control schedule and the groups that specifically pertain to the `func.call` lowering is necessary. All of the other stuff should (hopefully) be captured by other tests, so there's no reason to deliberately give ourself a larger maintenance burden.

Reply:
+1